Packages/CandidateUI/Sources/CandidateUI/VerticalCandidateController.swift

```diff
@@ -0,0 +1,556 @@
+        tableView.target = self
+
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(boundsChange),
+                                               name: NSView.boundsDidChangeNotification,
+                                               object: scrollView.contentView)
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    public override func reloadData() {
+        maxCandidateAttrStringWidth = ceil(candidateFont.pointSize + candidateTextPadding)
+        tableView.reloadData()
+        layoutCandidateView()
+        if delegate?.candidateCountForController(self) ?? 0 > 0 {
+            selectedCandidateIndex = 0
+        }
+    }
+
+    public override func showNextPage() -> Bool {
+        scrollPageByOne(true)
+    }
+
+    public override func showPreviousPage() -> Bool {
+        scrollPageByOne(false)
+    }
+
+    public override func highlightNextCandidate() -> Bool {
+        moveSelectionByOne(true)
+    }
+
+    public override func highlightPreviousCandidate() -> Bool {
+        moveSelectionByOne(false)
+    }
+
+    public override func candidateIndexAtKeyLabelIndex(_ index: UInt) -> UInt {
+        guard let delegate = delegate else {
+            return UInt.max
+        }
+
+        let firstVisibleRow = tableView.row(at: scrollView.documentVisibleRect.origin)
+        if firstVisibleRow != -1 {
+            let result = UInt(firstVisibleRow) + index
+            if result < delegate.candidateCountForController(self) {
+                return result
+            }
+        }
+
+        return UInt.max
+    }
+
+    public override var selectedCandidateIndex: UInt {
+        get {
+            let selectedRow = tableView.selectedRow
+            return selectedRow == -1 ? UInt.max : UInt(selectedRow)
+
+        }
+        set {
+            guard let delegate = delegate else {
+                return
+            }
+            var newIndex = newValue
+            let selectedRow = tableView.selectedRow
+            let labelCount = keyLabels.count
+            let itemCount = delegate.candidateCountForController(self)
+
+            if newIndex == UInt.max {
+                if itemCount == 0 {
+                    tableView.deselectAll(self)
+                    return
+                }
+                newIndex = 0
+            }
+
+            var lastVisibleRow = newValue
+
+            if selectedRow != -1 && itemCount > 0 && itemCount > labelCount {
+                if newIndex > selectedRow && (Int(newIndex) - selectedRow) > 1 {
+                    lastVisibleRow = min(newIndex + UInt(labelCount) - 1, itemCount - 1)
+                }
+                // no need to handle the backward case: (newIndex < selectedRow && selectedRow - newIndex > 1)
+            }
+
+            if itemCount > labelCount {
+                tableView.scrollRowToVisible(Int(lastVisibleRow))
```

Comment:
I ran into a crash here due to `lastVisibleRow` was still `UInt.max`, which in turn caused integer overflow...

It's unfortunate that we still have to work with Cocoa APIs that use `-1`. Should we revert all `UInt` uses back to `Int`, or should we at least use a wrapper function to make sure UInt->Int conversions are safe?

Reply:
Okay, I will fix that.